### ISSUE TITLE
BucketNotFoundException when using Predictive Optimization

### ISSUE THREAD

**Issue by davischmitz** [bucket4j/bucket4j#398]
Hi, 

I'm using ProxyManager with the Postgres integration to build the buckets and it was working just fine together with the spring boot tests of the application.

Due to performance issues, we decided to add the Prediction Optimization when building the bucket with ProxyManager. When the Optimization is enabled, I get several BucketNotFound exceptions during Spring Boot tests.

I tried adding `.withRecoveryStrategy(RecoveryStrategy.RECONSTRUCT)` but it didn't solve the issue.

Comparing DefaultBucketProxy.java and PredictiveCommandExecutor.java I see that the Predictive doesn't have this check that could avoid the error being thrown when using `RecoveryStrategy.RECONSTRUCT`: 

```
// the bucket was removed or lost, or not initialized yet, or needs to upgrade configuration
        if (result.isBucketNotFound() && recoveryStrategy == RecoveryStrategy.THROW_BUCKET_NOT_FOUND_EXCEPTION && wasInitializedBeforeExecution) {
            throw new BucketNotFoundException();
        }
```

Any suggestions on how to proceed?

**Comment by vladimir-bukhtoyarov** [comment]
@davischmitz please provide stacktrace. It looks as bug that should be fixed.

**Comment by davischmitz** [comment]
Hi @vladimir-bukhtoyarov , here is the stacktrace with the error:
```
Caused by: io.github.bucket4j.distributed.proxy.BucketNotFoundException: Bucket does not exist
	at io.github.bucket4j.distributed.remote.BucketNotFoundError.asException(BucketNotFoundError.java:43) ~[bucket4j-core-8.3.0.jar:8.3.0]
	at io.github.bucket4j.distributed.remote.CommandResult.getData(CommandResult.java:163) ~[bucket4j-core-8.3.0.jar:8.3.0]
	at io.github.bucket4j.distributed.proxy.optimization.predictive.PredictiveCommandExecutor.execute(PredictiveCommandExecutor.java:80) ~[bucket4j-core-8.3.0.jar:8.3.0]
	at io.github.bucket4j.distributed.proxy.optimization.batch.BatchingExecutor$3.apply(BatchingExecutor.java:58) ~[bucket4j-core-8.3.0.jar:8.3.0]
	at io.github.bucket4j.distributed.proxy.optimization.batch.BatchingExecutor$3.apply(BatchingExecutor.java:55) ~[bucket4j-core-8.3.0.jar:8.3.0]
	at io.github.bucket4j.util.concurrent.batch.BatchHelper.execute(BatchHelper.java:93) ~[bucket4j-core-8.3.0.jar:8.3.0]
	at io.github.bucket4j.distributed.proxy.optimization.batch.BatchingExecutor.execute(BatchingExecutor.java:77) ~[bucket4j-core-8.3.0.jar:8.3.0]
	at io.github.bucket4j.distributed.proxy.DefaultBucketProxy.execute(DefaultBucketProxy.java:208) ~[bucket4j-core-8.3.0.jar:8.3.0]
	at io.github.bucket4j.distributed.proxy.DefaultBucketProxy.tryConsumeAndReturnRemainingTokensImpl(DefaultBucketProxy.java:88) ~[bucket4j-core-8.3.0.jar:8.3.0]
	at io.github.bucket4j.AbstractBucket.tryConsumeAndReturnRemaining(AbstractBucket.java:328) ~[bucket4j-core-8.3.0.jar:8.3.0]
```

I confirmed that it happens both with the Predictive and the Delayed Optimizations.

**Comment by vladimir-bukhtoyarov** [comment]
@davischmitz 

Thank you. Problem is clear, I am going to provide fix at weekend.

**Comment by vladimir-bukhtoyarov** [comment]
@davischmitz 

The fix has been released with version ```8.4.RC1```.

Final release ```8.4.0``` will be available a few weeks later.

**Comment by vladimir-bukhtoyarov** [comment]
@davischmitz 

Also I do not reccomend to use following factory methods:
* https://github.com/bucket4j/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/optimization/Optimizations.java#L74
* https://github.com/bucket4j/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/optimization/Optimizations.java#L88

It configure wrong time precision. You need to use constructor directly like:
```java
new PredictiveOptimization(predictionParameters, delayParameters, NopeOptimizationListener.INSTANCE, TimeMeter.SYSTEM_MILLISECONDS);
```

**Comment by davischmitz** [comment]
Thanks @vladimir-bukhtoyarov! The new `8.4.RC1` version fixed our issue. 

Quick question about optimizations: 
- Can we use more than one at once? For example, use `batching + delay`

**Comment by vladimir-bukhtoyarov** [comment]
> Can we use more than one at once? For example, use batching + delay
No.

But fortunatelly, both delaying and predictive optimizations are already based on top of batching optimization. So when you configure dealying optimization you get the both batching and delaying.

**Comment by vladimir-bukhtoyarov** [comment]
released with version 8.4.0
